chia/full_node/mempool_manager.py

```diff
@@ -80,6 +93,84 @@ def validate_clvm_and_signature(
+    item: MempoolItem, dedup_coin_spends: Dict[bytes32, DedupCoinSpend]
+) -> Optional[Tuple[List[CoinSpend], uint64, Dict[bytes32, DedupCoinSpend], Set[Coin]]]:
+    consider_this_item = True
+    cost_saving = uint64(0)
+    spends_to_dedup: List[CoinSpend] = []
+    dedup_additions: Set[Coin] = set()
+    new_dedups: Dict[bytes32, DedupCoinSpend] = {}
+    # See if this item has coin spends that are eligible for deduplication
+    assert item.npc_result.conds is not None
+    eligible_coin_ids = [bytes32(s.coin_id) for s in item.npc_result.conds.spends if s.flags & ELIGIBLE_FOR_DEDUP]
+    for coin_id in eligible_coin_ids:
+        # Pick the first coin spend with coin_id
+        # npc_result.conds.spends should be consistent with
+        # spend_bundle.coin_spends at this point, guaranteeing the presence of
+        # this coin spend in this bundle as a result
+        coin_spend_in_bundle = next(cs for cs in item.spend_bundle.coin_spends if cs.coin.name() == coin_id)
+        solution_in_bundle = coin_spend_in_bundle.solution
+        # See if we processed an item with this coin before
+        if coin_id in dedup_coin_spends:
+            # See if the solution was identical
+            dedup_coin_spend = dedup_coin_spends[coin_id]
+            processed_solution = dedup_coin_spend.solution
+            duplicate_cost = dedup_coin_spend.cost
+            duplicate_additions = dedup_coin_spend.additions
+            if processed_solution == solution_in_bundle:
+                # Let's calculate the saved cost if we never did that before
+                if duplicate_cost is None:
+                    # If we never ran this before, additions should be empty
+                    assert len(duplicate_additions) == 0
+                    result = run_for_cost_and_additions(
+                        coin_id, coin_spend_in_bundle.puzzle_reveal, coin_spend_in_bundle.solution, item.cost
+                    )
+                    if result is None:
+                        # We failed to run this puzzle, skip this whole item
+                        consider_this_item = False
+                        cost_saving = uint64(0)
+                        break
+                    else:
+                        spend_cost, created_coins = result
+                        duplicate_cost = spend_cost
+                        duplicate_additions.update(created_coins)
+                        # If we end up including this item, update this entry's cost and additions
+                        new_dedups[coin_id] = DedupCoinSpend(processed_solution, spend_cost, duplicate_additions)
+                cost_saving = uint64(cost_saving + duplicate_cost)
+                dedup_additions.update(duplicate_additions)
+                spends_to_dedup.append(coin_spend_in_bundle)
+            else:
+                # It wasn't, so let's skip this whole transaction
+                consider_this_item = False
+                cost_saving = uint64(0)
+                break
```

Comment:
same thing here; you can just return `None`, no need to assign to the variables. I would also recommend inverting the condition in the if-statement to make this an early exit. Then you can reduce the level of indentation by one of the remaining code.